### PR TITLE
ISSUE-18 Protect BookingClerks against invalid requests

### PR DESCRIPTION
**Pull request by andkantor**
Validated scenarios:
1. No positions are requested
2. Already booked/payed position is requested
3. TicketRepository does not have the requested position

### CODE REVIEW

**Comment by andkantor** [comment]
@bearcheese But why only that one class?

**Comment by bearcheese** [comment]
Because other immutable classes covered by the filter? :)

File: adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/validator/BookingRequestValidator.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+        return VALID_REQUEST;
+    }
+
+    private boolean hasAlreadyBookedPosition(BookingRequest request) {
+        List<Position> bookedPositions = request.requestedPositions().stream()
+                .filter(position -> ticketRepository.get(position).status() != FREE)
+                .collect(toList());
+
+        return !bookedPositions.isEmpty();
```
Comment (bearcheese) [inline adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/validator/BookingRequestValidator.java]: What about:
```
return request.requestedPositions().stream()
                .anyMatch(position -> ticketRepository.get(position).status() != FREE);
```

File: adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/BookingRequest.java [hunk 2]
```diff
@@ -13,6 +19,16 @@
...
     List<Integer> positions();
 
+    default ImmutableList<Position> requestedPositions() {
+        return positions().stream()
+                .map(positionId -> Position.builder()
+                        .eventId(eventId())
+                        .sectorId(sectorId())
+                        .id(positionId)
+                        .build())
+                .collect(collectingAndThen(toList(), ImmutableList::copyOf));
```
Comment (bearcheese) [inline adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/BookingRequest.java]: Revelation of the day: Guava 21.0 `ImmutableList` has a method called `toImmutableList()`, therefore the above line could be simplified to `.collect(ImmutableList.toImmutableList());`
Side note: this is a good candidate for a KAIZEN.

File: adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/TicketRepository.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package com.microservicesteam.adele.booking.domain;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.google.common.collect.ImmutableList;
+import com.microservicesteam.adele.ticketmaster.model.Position;
+import com.microservicesteam.adele.ticketmaster.model.Ticket;
+
+public class TicketRepository {
```
Comment (bearcheese) [inline adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/TicketRepository.java]: 👍

**Comment by bearcheese** [comment]
One more thing: I would put some tests into `BookingServiceTest` to see validation logic is working, so add at least one not valid scenario.

File: adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/validator/ValidationResult.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package com.microservicesteam.adele.booking.domain.validator;
+
+public enum ValidationResult {
+
+    VALID_REQUEST(true, "000", "Valid request"),
+    INVALID_POSITIONS_EMPTY(false, "001", "At least one position must be requested"),
+    INVALID_POSITIONS_BOOKED(false, "002", "At least one of the positions is already booked");
+
+    private boolean valid;
```
Comment (kelemena) [inline adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/validator/ValidationResult.java]: Are we against `public final`?

Comment (andkantor) [inline adele-booking/src/main/java/com/microservicesteam/adele/booking/domain/validator/ValidationResult.java]: Not against them, but better like hiding internal representation.

**Comment by andkantor** [comment]
@bearcheese @kelemena addressed your comments, please review.

**Pull request merged**